File: ts/src/lib/emitters/morph-shape.ts
import { WritableBitStream, WritableByteStream, WritableStream } from "@open-flash/stream";
import incident from "incident";
import { Uint2, Uint5, Uint8, Uint16, UintSize } from "semantic-types";
import { FillStyleType } from "swf-types";
import * as fillStyles from "swf-types/fill-styles/index";
import { JoinStyleType } from "swf-types/join-styles/_type";
import { MorphFillStyle } from "swf-types/morph-fill-style";
import { MorphLineStyle } from "swf-types/morph-line-style";
import { MorphShapeRecord } from "swf-types/morph-shape-record";
import { MorphShapeStyles } from "swf-types/morph-shape-styles";
import { MorphShape } from "swf-types/morph-shape";
import { ShapeRecordType } from "swf-types/shape-records/_type";
import { MorphStyleChange } from "swf-types/shape-records/morph-style-change";

import { getSintMinBitCount, getUintBitCount } from "../get-bit-count.js";
import { emitMatrix, emitStraightSRgba8 } from "./basic-data-types.js";
import { emitMorphGradient } from "./gradient.js";
import { capStyleToCode, emitEdgeBits, emitListLength, joinStyleToCode } from "./shape.js";

export enum MorphShapeVersion {
  MorphShape1 = 1,
  MorphShape2 = 2,
}

export function emitMorphShape(
  byteStream: WritableByteStream,
  value: MorphShape,
  morphShapeVersion: MorphShapeVersion,
): void {
  const shapeStream: WritableStream = new WritableStream();
  const startShapeSize: UintSize = emitMorphShapeBits(shapeStream, value, morphShapeVersion);
  byteStream.writeUint32LE(startShapeSize);
  byteStream.write(shapeStream);
}

/**
 * @param bitStream Bitstream where the morph shape should be emitted
 * @param value Morph shape to emit
 * @param morphShapeVersion Morphshape version to use
 * @return Size of the start shape in bytes
 */
export function emitMorphShapeBits(
  bitStream: WritableBitStream,
  value: MorphShape,
  morphShapeVersion: MorphShapeVersion,
): UintSize {
  const [fillBits, lineBits] = emitMorphShapeStylesBits(bitStream, value.initialStyles, morphShapeVersion);
  emitMorphShapeStartRecordStringBits(bitStream, value.records, fillBits, lineBits, morphShapeVersion);
  bitStream.align();

  const result: UintSize = bitStream.bytePos;

  // `0` for style bits: there are no style changes in the end state.
  bitStream.writeUint32Bits(4, 0);
  bitStream.writeUint32Bits(4, 0);
  emitMorphShapeEndRecordStringBits(bitStream, value.records);
  bitStream.align();

  return result;
}

/**
 *
 * @return [fillBits, lineBits]
 */
export function emitMorphShapeStylesBits(
  bitStream: WritableBitStream,
  value: MorphShapeStyles,
  morphShapeVersion: MorphShapeVersion,
): [UintSize, UintSize] {
  const byteStream: WritableByteStream = bitStream.asByteStream();
  emitMorphFillStyleList(byteStream, value.fill);
  emitMorphLineStyleList(byteStream, value.line, morphShapeVersion);
  const fillBits: UintSize = getUintBitCount(value.fill.length);
  const lineBits: UintSize = getUintBitCount(value.line.length);
  bitStream.writeUint32Bits(4, fillBits);
  bitStream.writeUint32Bits(4, lineBits);
  return [fillBits, lineBits];
}

export function emitMorphShapeStartRecordStringBits(
  bitStream: WritableBitStream,
  value: MorphShapeRecord[],
  fillBits: UintSize,
  lineBits: UintSize,
  morphShapeVersion: MorphShapeVersion,
): void {
  for (const record of value) {
    if (record.type === ShapeRecordType.StyleChange) {
      bitStream.writeBoolBits(false); // isEdge
      [fillBits, lineBits] = emitMorphStyleChangeBits(bitStream, record, fillBits, lineBits, morphShapeVersion);
    } else {
      // assert(record.type === ShapeRecordType.Edge)
      bitStream.writeBoolBits(true); // isEdge
      emitEdgeBits(bitStream, record);
    }
  }
  bitStream.writeUint16Bits(6, 0);
}

export function emitMorphShapeEndRecordStringBits(
  bitStream: WritableBitStream,
  value: MorphShapeRecord[],
): void {
  for (const record of value) {
    if (record.type === ShapeRecordType.StyleChange) {
      if (record.moveTo === undefined) {
        continue;
      }
      bitStream.writeBoolBits(false); // isEdge
      const flags: Uint5 = 0b00001; // Pure `moveTo`
      bitStream.writeUint16Bits(5, flags);
      if (record.morphMoveTo === undefined) {
        throw new incident.Incident("UndefinedEndMoveTo");
      }
      const bitCount: UintSize = getSintMinBitCount(record.morphMoveTo.x, record.morphMoveTo.y);
      bitStream.writeUint16Bits(5, bitCount);
      bitStream.writeSint32Bits(bitCount, record.morphMoveTo.x);
      bitStream.writeSint32Bits(bitCount, record.morphMoveTo.y);
    } else {
      // assert(record.type === ShapeRecordType.Edge)
      bitStream.writeBoolBits(true); // isEdge
      emitEdgeBits(
        bitStream,
        {
          type: ShapeRecordType.Edge,
          delta: record.morphDelta,
          controlDelta: record.morphControlDelta,
        },
      );
    }
  }
  bitStream.writeUint16Bits(6, 0);
}

export function emitMorphStyleChangeBits(
  bitStream: WritableBitStream,
  value: MorphStyleChange,
  fillBits: UintSize,
  lineBits: UintSize,
  morphShapeVersion: MorphShapeVersion,
): [UintSize, UintSize] {
  const hasMoveTo: boolean = value.moveTo !== undefined;
  const hasNewLeftFill: boolean = value.leftFill !== undefined;
  const hasNewRightFill: boolean = value.rightFill !== undefined;
  const hasNewLineStyle: boolean = value.lineStyle !== undefined;
  const hasNewStyles: boolean = value.newStyles !== undefined;

  const flags: Uint5 = 0
    | (hasMoveTo ? 1 << 0 : 0)
    | (hasNewLeftFill ? 1 << 1 : 0)
    | (hasNewRightFill ? 1 << 2 : 0)
    | (hasNewLineStyle ? 1 << 3 : 0)
    | (hasNewStyles ? 1 << 4 : 0);

  bitStream.writeUint32Bits(5, flags);

  if (hasMoveTo) {
    const bitCount: UintSize = getSintMinBitCount(value.moveTo!.x, value.moveTo!.y);
    bitStream.writeUint16Bits(5, bitCount);
    bitStream.writeSint32Bits(bitCount, value.moveTo!.x);
    bitStream.writeSint32Bits(bitCount, value.moveTo!.y);
  }

  if (hasNewLeftFill) {
    bitStream.writeUint16Bits(fillBits, value.leftFill!);
  }
  if (hasNewRightFill) {
    bitStream.writeUint16Bits(fillBits, value.rightFill!);
  }
  if (hasNewLineStyle) {
    bitStream.writeUint16Bits(lineBits, value.lineStyle!);
  }

  if (hasNewStyles) {
    [fillBits, lineBits] = emitMorphShapeStylesBits(bitStream, value.newStyles!, morphShapeVersion);
  }

  return [fillBits, lineBits];
}

export function emitMorphFillStyleList(byteStream: WritableByteStream, value: MorphFillStyle[]): void {
  emitListLength(byteStream, value.length, true);
  for (const fillStyle of value) {
    emitMorphFillStyle(byteStream, fillStyle);
  }
}

export function emitMorphFillStyle(byteStream: WritableByteStream, value: MorphFillStyle): void {
  switch (value.type) {
    case FillStyleType.Bitmap: {
      const code: Uint8 = 0
        | (!value.repeating ? 1 << 0 : 0)
        | (!value.smoothed ? 1 << 1 : 0)
        | 0x40;
      byteStream.writeUint8(code);
      emitMorphBitmapFill(byteStream, value);
      break;
    }
    case FillStyleType.FocalGradient: {
      byteStream.writeUint8(0x13);
      emitMorphFocalGradientFill(byteStream, value);
      break;
    }
    case FillStyleType.LinearGradient: {
      byteStream.writeUint8(0x10);
      emitMorphLinearGradientFill(byteStream, value);
      break;
    }
    case FillStyleType.RadialGradient: {
      byteStream.writeUint8(0x12);
      emitMorphRadialGradientFill(byteStream, value);
      break;
    }
    case FillStyleType.Solid: {
      byteStream.writeUint8(0x00);
      emitMorphSolidFill(byteStream, value);
      break;
    }
    default: {
      throw new incident.Incident("UnexpectedMorphFillStyle");
    }
  }
}

export function emitMorphBitmapFill(byteStream: WritableByteStream, value: fillStyles.MorphBitmap): void {
  byteStream.writeUint16LE(value.bitmapId);
  emitMatrix(byteStream, value.matrix);
  emitMatrix(byteStream, value.morphMatrix);
}

export function emitMorphFocalGradientFill(byteStream: WritableByteStream, value: fillStyles.MorphFocalGradient): void {
  emitMatrix(byteStream, value.matrix);
  emitMatrix(byteStream, value.morphMatrix);
  emitMorphGradient(byteStream, value.gradient);
  byteStream.writeSint16LE(value.focalPoint.epsilons);
  byteStream.writeSint16LE(value.morphFocalPoint.epsilons);
}

export function emitMorphLinearGradientFill(
  byteStream: WritableByteStream,
  value: fillStyles.MorphLinearGradient,
): void {
  emitMatrix(byteStream, value.matrix);
  emitMatrix(byteStream, value.morphMatrix);
  emitMorphGradient(byteStream, value.gradient);
}

export function emitMorphRadialGradientFill(
  byteStream: WritableByteStream,
  value: fillStyles.MorphRadialGradient,
): void {
  emitMatrix(byteStream, value.matrix);
  emitMatrix(byteStream, value.morphMatrix);
  emitMorphGradient(byteStream, value.gradient);
}

export function emitMorphSolidFill(byteStream: WritableByteStream, value: fillStyles.MorphSolid): void {
  emitStraightSRgba8(byteStream, value.color);
  emitStraightSRgba8(byteStream, value.morphColor);
}

export function emitMorphLineStyleList(
  byteStream: WritableByteStream,
  value: MorphLineStyle[],
  morphShapeVersion: MorphShapeVersion,
): void {
  emitListLength(byteStream, value.length, true);
  for (const lineStyle of value) {
    if (morphShapeVersion < MorphShapeVersion.MorphShape2) {
      emitMorphLineStyle1(byteStream, lineStyle);
    } else {
      emitMorphLineStyle2(byteStream, lineStyle);
    }
  }
}

export function emitMorphLineStyle1(byteStream: WritableByteStream, value: MorphLineStyle): void {
  if (value.fill.type !== FillStyleType.Solid) {
    throw new incident.Incident("ExpectedSolidMorphFill");
  }
  byteStream.writeUint16LE(value.width);
  byteStream.writeUint16LE(value.morphWidth);
  emitMorphSolidFill(byteStream, value.fill);
}

export function emitMorphLineStyle2(byteStream: WritableByteStream, value: MorphLineStyle): void {
  byteStream.writeUint16LE(value.width);
  byteStream.writeUint16LE(value.morphWidth);

  const hasFill: boolean = value.fill.type !== FillStyleType.Solid;
  const joinStyleCode: Uint2 = joinStyleToCode(value.join.type);
  const startCapStyleCode: Uint2 = capStyleToCode(value.startCap);
  const endCapStyleCode: Uint2 = capStyleToCode(value.endCap);

  const flags: Uint16 = 0
    | (value.pixelHinting ? 1 << 0 : 0)
    | (value.noVScale ? 1 << 1 : 0)
    | (value.noHScale ? 1 << 2 : 0)
    | (hasFill ? 1 << 3 : 0)
    | ((joinStyleCode & 0b11) << 4)
    | ((startCapStyleCode & 0b11) << 6)
    | ((endCapStyleCode & 0b11) << 8)
    | (value.noClose ? 1 << 10 : 0);
  // Skip bits [11, 15]
  byteStream.writeUint16LE(flags);

  if (value.join.type === JoinStyleType.Miter) {
    byteStream.writeSint16LE(value.join.limit.epsilons);
  }

  if (hasFill) {
    emitMorphFillStyle(byteStream, value.fill);
  } else {
    emitMorphSolidFill(byteStream, value.fill as fillStyles.MorphSolid);
  }
}

File: rs/src/morph_shape.rs
use std::convert::{TryFrom, TryInto};
use std::io;

use swf_tree as ast;

use crate::basic_data_types::{emit_matrix, emit_straight_s_rgba8};
use crate::bit_count::{get_i32_min_bit_count, get_u32_bit_count};
use crate::gradient::emit_morph_gradient;
use crate::io_bits::{BitsWriter, WriteBits};
use crate::primitives::{emit_le_i16, emit_le_u16, emit_le_u32, emit_u8};
use crate::shape::{cap_style_to_code, emit_edge_bits, emit_list_length, join_style_to_code};

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum MorphShapeVersion {
  MorphShape1,
  MorphShape2,
}

pub(crate) fn emit_morph_shape<W: io::Write>(writer: &mut W, value: &ast::MorphShape, version: MorphShapeVersion) -> io::Result<()> {
  let mut bits_writer = BitsWriter::new(Vec::new());
  let (fill_bits, line_bits) = emit_morph_shape_styles_bits(&mut bits_writer, &value.initial_styles, version)?;
  emit_morph_shape_start_record_string_bits(
    &mut bits_writer,
    &value.records,
    fill_bits,
    line_bits,
    version,
  )?;
  let inner_bits_writer = bits_writer.into_inner()?;
  let start_size = inner_bits_writer.len();

  let mut bits_writer = BitsWriter::new(inner_bits_writer);

  // TODO: We should be able to skip these bits (no styles used for the endRecords)
  // We copy the bits from the start shape to match the behavior in `morph-rotating-square`.
  bits_writer.write_u32_bits(4, fill_bits)?;
  bits_writer.write_u32_bits(4, line_bits)?;
  emit_morph_shape_end_record_string_bits(&mut bits_writer,&value.records)?;

  emit_le_u32(writer, start_size.try_into().unwrap())?;
  writer.write_all(&bits_writer.into_inner()?)
}

pub(crate) fn emit_morph_shape_styles_bits<W: WriteBits>(writer: &mut W, value: &ast::MorphShapeStyles, version: MorphShapeVersion) -> io::Result<(u32, u32)> {
  let bytes_writer = writer.write_bytes()?;
  emit_morph_fill_style_list(bytes_writer, &value.fill)?;
  emit_morph_line_style_list(bytes_writer, &value.line, version)?;
  // The max style `id` is `.len()` (and not `.len() - 1`) because `0` always
  // represents the empty style and custom styles are 1-indexed.
  let max_fill_id: u32 = u32::try_from(value.fill.len()).unwrap();
  let max_line_id: u32 = u32::try_from(value.line.len()).unwrap();
  let fill_bits: u32 = get_u32_bit_count(max_fill_id);
  let line_bits: u32 = get_u32_bit_count(max_line_id);
  writer.write_u32_bits(4, fill_bits)?;
  writer.write_u32_bits(4, line_bits)?;
  Ok((fill_bits, line_bits))
}

pub(crate) fn emit_morph_shape_start_record_string_bits<W: WriteBits>(writer: &mut W, value: &[ast::MorphShapeRecord], mut fill_bits: u32, mut line_bits: u32, version: MorphShapeVersion) -> io::Result<()> {
  for record in value {
    match record {
      ast::MorphShapeRecord::Edge(ref record) => {
        writer.write_bool_bits(true)?; // is_edge
        emit_edge_bits(
          writer,
          &ast::shape_records::Edge { delta: record.delta, control_delta: record.control_delta },
        )?;
      }
      ast::MorphShapeRecord::StyleChange(ref record) => {
        writer.write_bool_bits(false)?; // is_edge
        let (next_fill_bits, next_line_bits) = emit_morph_style_change_bits(writer, record, fill_bits, line_bits, version)?;
        fill_bits = next_fill_bits;
        line_bits = next_line_bits;
      }
    }
  }
  writer.write_u32_bits(6, 0)
}

pub(crate) fn emit_morph_shape_end_record_string_bits<W: WriteBits>(writer: &mut W, value: &[ast::MorphShapeRecord]) -> io::Result<()> {
  for record in value {
    match record {
      ast::MorphShapeRecord::Edge(ref record) => {
        writer.write_bool_bits(true)?; // is_edge
        emit_edge_bits(
          writer,
          &ast::shape_records::Edge { delta: record.morph_delta, control_delta: record.morph_control_delta },
        )?;
      }
      ast::MorphShapeRecord::StyleChange(ref record) => {
        if record.move_to.is_none() {
          continue;
        }
        writer.write_bool_bits(false)?; // is_edge
        let flags: u8 = 0b00001; // Pure `moveTo`
        writer.write_u32_bits(5, flags.into())?;
        let morph_move_to = record.morph_move_to.unwrap();
        let bits = get_i32_min_bit_count(vec![morph_move_to.x, morph_move_to.y].into_iter());
        writer.write_u32_bits(5, bits)?;
        writer.write_i32_bits(bits, morph_move_to.x)?;
        writer.write_i32_bits(bits, morph_move_to.y)?;
      }
    }
  }
  writer.write_u32_bits(6, 0)
}

pub(crate) fn emit_morph_style_change_bits<W: WriteBits>(
  writer: &mut W,
  value: &ast::shape_records::MorphStyleChange,
  fill_bits: u32,
  line_bits: u32,
  version: MorphShapeVersion,
) -> io::Result<(u32, u32)> {
  let has_move_to = value.move_to.is_some();
  let has_new_left_fill = value.left_fill.is_some();
  let has_new_right_fill = value.right_fill.is_some();
  let has_new_line_style = value.line_style.is_some();
  let has_new_styles = value.new_styles.is_some();

  let flags: u8 = 0
    | (if has_move_to { 1 << 0 } else { 0 })
    | (if has_new_left_fill { 1 << 1 } else { 0 })
    | (if has_new_right_fill { 1 << 2 } else { 0 })
    | (if has_new_line_style { 1 << 3 } else { 0 })
    | (if has_new_styles { 1 << 4 } else { 0 });

  assert_ne!(flags, 0);

  writer.write_u32_bits(5, flags.into())?;

  if let Some(move_to) = value.move_to {
    let bits = get_i32_min_bit_count(vec![move_to.x, move_to.y].into_iter());
    writer.write_u32_bits(5, bits)?;
    writer.write_i32_bits(bits, move_to.x)?;
    writer.write_i32_bits(bits, move_to.y)?;
  }

  if let Some(left_fill) = value.left_fill {
    writer.write_u32_bits(fill_bits, left_fill.try_into().unwrap())?;
  }
  if let Some(right_fill) = value.right_fill {
    writer.write_u32_bits(fill_bits, right_fill.try_into().unwrap())?;
  }
  if let Some(line_style) = value.line_style {
    writer.write_u32_bits(line_bits, line_style.try_into().unwrap())?;
  }

  if let Some(ref new_styles) = &value.new_styles {
    emit_morph_shape_styles_bits(writer, new_styles, version)
  } else {
    Ok((fill_bits, line_bits))
  }
}

pub(crate) fn emit_morph_fill_style_list<W: io::Write + ?Sized>(writer: &mut W, value: &[ast::MorphFillStyle]) -> io::Result<()> {
  emit_list_length(writer, value.len(), true)?;
  for fill_style in value {
    emit_morph_fill_style(writer, fill_style)?;
  }
  Ok(())
}

pub(crate) fn emit_morph_fill_style<W: io::Write + ?Sized>(writer: &mut W, value: &ast::MorphFillStyle) -> io::Result<()> {
  match value {
    ast::MorphFillStyle::Bitmap(ref style) => {
      let code: u8 = 0
        | (if !style.repeating { 1 << 0 } else { 0 })
        | (if !style.smoothed { 1 << 1 } else { 0 })
        | 0x40;
      emit_u8(writer, code)?;
      emit_morph_bitmap_fill(writer, style)
    }
    ast::MorphFillStyle::FocalGradient(ref style) => {
      emit_u8(writer, 0x13)?;
      emit_morph_focal_gradient_fill(writer, style)
    }
    ast::MorphFillStyle::LinearGradient(ref style) => {
      emit_u8(writer, 0x10)?;
      emit_morph_linear_gradient_fill(writer, style)
    }
    ast::MorphFillStyle::RadialGradient(ref style) => {
      emit_u8(writer, 0x12)?;
      emit_morph_radial_gradient_fill(writer, style)
    }
    ast::MorphFillStyle::Solid(ref style) => {
      emit_u8(writer, 0x00)?;
      emit_morph_solid_fill(writer, style)
    }
  }
}

pub(crate) fn emit_morph_bitmap_fill<W: io::Write + ?Sized>(writer: &mut W, value: &ast::fill_styles::MorphBitmap) -> io::Result<()> {
  emit_le_u16(writer, value.bitmap_id)?;
  emit_matrix(writer, &value.matrix)?;
  emit_matrix(writer, &value.morph_matrix)
}

pub(crate) fn emit_morph_focal_gradient_fill<W: io::Write + ?Sized>(writer: &mut W, value: &ast::fill_styles::MorphFocalGradient) -> io::Result<()> {
  emit_matrix(writer, &value.matrix)?;
  emit_matrix(writer, &value.morph_matrix)?;
  emit_morph_gradient(writer, &value.gradient)?;
  emit_le_i16(writer, value.focal_point.epsilons)?;
  emit_le_i16(writer, value.morph_focal_point.epsilons)
}

pub(crate) fn emit_morph_linear_gradient_fill<W: io::Write + ?Sized>(writer: &mut W, value: &ast::fill_styles::MorphLinearGradient) -> io::Result<()> {
  emit_matrix(writer, &value.matrix)?;
  emit_matrix(writer, &value.morph_matrix)?;
  emit_morph_gradient(writer, &value.gradient)
}

pub(crate) fn emit_morph_radial_gradient_fill<W: io::Write + ?Sized>(writer: &mut W, value: &ast::fill_styles::MorphRadialGradient) -> io::Result<()> {
  emit_matrix(writer, &value.matrix)?;
  emit_matrix(writer, &value.morph_matrix)?;
  emit_morph_gradient(writer, &value.gradient)
}

pub(crate) fn emit_morph_solid_fill<W: io::Write + ?Sized>(writer: &mut W, value: &ast::fill_styles::MorphSolid) -> io::Result<()> {
  emit_straight_s_rgba8(writer, value.color)?;
  emit_straight_s_rgba8(writer, value.morph_color)
}

pub(crate) fn emit_morph_line_style_list<W: io::Write + ?Sized>(writer: &mut W, value: &[ast::MorphLineStyle], version: MorphShapeVersion) -> io::Result<()> {
  emit_list_length(writer, value.len(), true)?;
  for line_style in value {
    if version < MorphShapeVersion::MorphShape2 {
      emit_morph_line_style1(writer, line_style)?;
    } else {
      emit_morph_line_style2(writer, line_style)?;
    }
  }
  Ok(())
}

pub(crate) fn emit_morph_line_style1<W: io::Write + ?Sized>(writer: &mut W, value: &ast::MorphLineStyle) -> io::Result<()> {
  match value.fill {
    ast::MorphFillStyle::Solid(ref style) => {
      emit_le_u16(writer, value.width)?;
      emit_le_u16(writer, value.morph_width)?;
      emit_morph_solid_fill(writer, style)
    }
    _ => panic!("InvalidMorphLineStyle1Fill")
  }
}

pub(crate) fn emit_morph_line_style2<W: io::Write + ?Sized>(writer: &mut W, value: &ast::MorphLineStyle) -> io::Result<()> {
  emit_le_u16(writer, value.width)?;
  emit_le_u16(writer, value.morph_width)?;

  let has_fill = match &value.fill {
    ast::MorphFillStyle::Solid(_) => false,
    _ => true
  };
  let join_style_code = join_style_to_code(value.join);
  let start_cap_style_code = cap_style_to_code(value.start_cap);
  let end_cap_style_code = cap_style_to_code(value.end_cap);

  let flags: u16 = 0
    | (if value.pixel_hinting { 1 << 0 } else { 0 })
    | (if value.no_v_scale { 1 << 1 } else { 0 })
    | (if value.no_h_scale { 1 << 2 } else { 0 })
    | (if has_fill { 1 << 3 } else { 0 })
    | ((u16::from(join_style_code) & 0b11) << 4)
    | ((u16::from(start_cap_style_code) & 0b11) << 6)
    | ((u16::from(end_cap_style_code) & 0b11) << 8)
    | (if value.no_close { 1 << 10 } else { 0 });
  // Skip bits [11, 15]
  emit_le_u16(writer, flags)?;

  if let ast::JoinStyle::Miter(miter) = value.join {
    emit_le_u16(writer, miter.limit)?;
  }

  match &value.fill {
    ast::MorphFillStyle::Solid(ref style) => emit_straight_s_rgba8(writer, style.color),
    style => emit_morph_fill_style(writer, style)
  }
}
